### DISCUSSION TITLE
How to support Python PEP 503 repos with custom URLs (not ending with /simple)

### DISCUSSION THREAD

**Discussion by maxromanovsky** [renovatebot/renovate#26967] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### What is your question?

Current logic of detecting PEP 503 repos is rather limited (however, it seems nobody complains about it):
https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/pypi/index.ts#L45
```python
if (hostUrl.endsWith('/simple/') || hostUrl.endsWith('/+simple/'))
```

It was introduced in #3125 and enhanced in #4508

The use case that is not currently covered is buf.build repos, that provide pre-generated SDKs for gRPC specs:
https://buf.build/docs/bsr/generated-sdks/python

Repo URL in this case is `https://buf.build/gen/python` (no `/simple` suffix)

Example of package page:
https://buf.build/gen/python/connectrpc-eliza-grpc-python

And Renovate tries to handle it as a PEP 691 (`/json`) repo.

I'd be happy to contribute, same as I did for #24581

I think adding some metadata into `hostRules` (and keeping backwards compatibility with #3125 & #4508) would be suitable in this case.

### Logs (if relevant)

<details><summary>Logs</summary>

```

Copy/paste the relevant log(s) here, between the starting and ending backticks

```

</details>

**Comment by maxromanovsky** [comment]
Upd: buf.build actually supports `/simple/` suffix, but something is not working for me anyways.
I've used Mend Renovate, tomorrow I'll try running it locally with debugging.

```
DEBUG: Failed to look up pypi package connectrpc-eliza-grpc-python
{
  "dependency": "connectrpc-eliza-grpc-python"
  "packageFile": "Pipfile"
}

WARN: Package lookup failures
{
  "warnings": [
    "Failed to look up pypi package connectrpc-eliza-grpc-python"
  ]
  "files": [
    "Pipfile"
  ]
}
```

**Comment by rarkins** [comment]
FYI this seems related: #26483

**Reply by maxromanovsky** [reply]
@rarkins I found an issue.
After fixing and running Renovate locally it created a PR: https://github.com/maxromanovsky/tmp-buf-pipenv/pull/4

The issue is with this line: https://github.com/renovatebot/renovate/blob/b69416ce1745f67c9fc1d149738e2f52feb4f732/lib/modules/datasource/pypi/index.ts#L249
After replacing it with `link.text.trim()` it started working.
I didn't test it for regressions yet.

The root cause what HTML formatting, where links looked like this:
```html
<a
             href="https://buf.build/gen/python/connectrpc-eliza-grpc-python/connectrpc_eliza_grpc_python-1.60.0.1.20230913231627&#43;233fca715f49-py3-none-any.whl"
             data-requires-python="&gt;=3.7"
             >
             connectrpc_eliza_grpc_python-1.60.0.1.20230913231627&#43;233fca715f49-py3-none-any.whl
           </a>
```

And then the extracted HTML node text looked like this:
```javascript
"\n      connectrpc_eliza_grpc_python-1.59.3.1.20230727061922+edc8ae343b96-py3-none-any.whl\n    "
```

**Reply by rarkins** [reply]
There should be nothing wrong with trimming, so please submit a PR when you can

**Comment by maxromanovsky** [comment]
I created a PR: https://github.com/renovatebot/renovate/pull/27013
